### PR TITLE
#1321 Transaction script model implementation

### PR DESCRIPTION
**Pull request by ashishtrivedi16**

### CODE REVIEW

**Comment by iluwatar** [comment]
@ashishtrivedi16  there is a conflict that must resolved

File: transaction-script/Readme.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
+---
+layout: pattern
+title: Transaction script
+folder: transaction-script
+permalink: /patterns/transaction-script/
+categories: Domain logic
```
Comment (iluwatar) [inline transaction-script/Readme.md:6]:
```suggestion
categories: Behavioral
```

File: transaction-script/Readme.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+tags:
+ - Data access
+---
+
+## Intent
+Transaction script(TS) is mainly used in small applications where nothing complex is done and bigger architecture's are not needed.
```
Comment (iluwatar) [inline transaction-script/Readme.md:12]:
```suggestion
Transaction Script organizes business logic by procedures where each procedure handles a single request from the presentation.
```

File: transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/CustomException.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.ashishtrivedi16.transactionscript;
+
+/**
+ * Custom exception.
+ */
+public class CustomException extends Exception {
```
Comment (iluwatar) [inline transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/CustomException.java:29]: Can we create a more descriptive class name here? The expection name should indicate what kind of error has occured.

File: transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/Hotel.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    Optional<Room> room = hotelDao.getById(roomNumber);
+
+    if (room.isEmpty()) {
+      throw new Exception("Room number: " + roomNumber + " does not exist");
+    } else {
+      if (room.get().isBooked()) {
+        throw new Exception("Room already booked!");
+      } else {
+        Room updateRoomBooking = room.get();
```
Comment (iluwatar) [inline transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/Hotel.java:55]: Please use Java's local variable type-inference `var` where possible

**Comment by iluwatar** [comment]
Please find the review comments above. When you're ready for another review please comment on this issue. Thanks!

**Comment by ashishtrivedi16** [comment]
Hey @iluwatar can you review the changes again?

File: transaction-script/Readme.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+In plain words
+> All logic related to booking a hotel room like checking room availability,
+> calculate rates and update the database is done inside a single transaction script.
+> Similar procedure is also needed for cancelling a room booking and all 
+> that logic will be in another transaction script.
+
+Programmatic example
+
+The Hotel class takes care of booking and cancelling a room in a hotel.
```
Comment (iluwatar) [inline transaction-script/Readme.md:27]:
```suggestion
The `Hotel` class takes care of booking and cancelling room reservations.
```

File: transaction-script/Readme.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        LOGGER.info(refundAmount + " is refunded");
+      } else {
+        throw new Exception("No booking for the room exists");
+      }
+    }
+  }
+}
+```
+
+This class has two methods, one for booking and cancelling a room respectively.
```
Comment (iluwatar) [inline transaction-script/Readme.md:79]:
```suggestion
The `Hotel` class has two methods, one for booking and cancelling a room respectively. Each one of them handles a single transaction in the system, making `Hotel` implement the Transaction Script pattern.
```

File: transaction-script/Readme.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+```
+The cancel room method consolidates steps like checking if the room is booked or not, 
+if booked then calculates the refund amount and updates the database using the DAO.
+
+## Class diagram
+![alt text](./etc/transaction-script.png "Transaction script model")
+
+## Applicability
+Use the transaction script model when the application has only a small amount of logic and that
+logic won't be extended in the future.
```
Comment (iluwatar) [inline transaction-script/Readme.md:98]:
```suggestion
Use the Transaction Script pattern when the application has only a small amount of logic and that logic won't be extended in the future.
```

File: transaction-script/Readme.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+
+## Known uses
+* Revenue recognition in business systems.
```
Comment (iluwatar) [inline transaction-script/Readme.md:102]: I think in this section we need to list very specific uses. So if we can't point to an implementation let's leave this section out.

File: transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/TransactionScriptApp.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
+/*
```
Comment (iluwatar) [inline transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/TransactionScriptApp.java:1]: Let's name the main class `App.java` so it's similar to other patterns

File: transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/TransactionScriptApp.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+package com.ashishtrivedi16.transactionscript;
+
+import java.util.List;
+import javax.sql.DataSource;
+import org.h2.jdbcx.JdbcDataSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TransactionScriptApp {
```
Comment (iluwatar) [inline transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/TransactionScriptApp.java:32]: Above the class please add comments how the example implements the pattern. See the other patterns for examples.

File: transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/TransactionScriptApp.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  private static final String H2_DB_URL = "jdbc:h2:~/test";
+  private static final Logger LOGGER = LoggerFactory.getLogger(TransactionScriptApp.class);
+
+  /**
+   * Program entry point.
+   *
+   * @param args command line arguments
+   * @throws Exception if any error occurs
+   */
+  public static void main(String[] args) throws Exception {
```
Comment (iluwatar) [inline transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/TransactionScriptApp.java:43]: Add comments that explain what is going on here. I see there's some initialization and then the calls to `hotel` variable are the important ones that trigger the transactions. Finally there's some cleanup.

File: transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/Hotel.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+package com.ashishtrivedi16.transactionscript;
+
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Hotel {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TransactionScriptApp.class);
+
+  private HotelDaoImpl hotelDao;
```
Comment (iluwatar) [inline transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/Hotel.java:33]: Should be `final`

File: transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/HotelDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+            return true;
+          } catch (Exception e) {
+            throw new RuntimeException(e); // NOSONAR
+          }
+        }
+      }, false).onClose(() -> {
+        try {
+          mutedClose(connection, statement, resultSet);
+        } catch (Exception e) {
+          e.printStackTrace();
```
Comment (iluwatar) [inline transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/HotelDaoImpl.java:70]: Use logger

File: transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/HotelDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+    try (var connection = getConnection();
+         var statement = connection.prepareStatement("INSERT INTO ROOMS VALUES (?,?,?,?)")) {
+      statement.setInt(1, room.getId());
+      statement.setString(2, room.getRoomType());
+      statement.setInt(3, room.getPrice());
+      statement.setBoolean(4, room.isBooked());
+      statement.execute();
+      return true;
+    } catch (Exception ex) {
+      throw new SqlException(ex.getMessage(), ex);
```
Comment (iluwatar) [inline transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/HotelDaoImpl.java:116]: Looks a bit suspicious. Why are we catching any exception and throwing `SqlException`? Is `SqlException` caught somewhere?

**Comment by iluwatar** [comment]
@ashishtrivedi16 let me know when it's ready for another review

**Comment by iluwatar** [comment]
@ashishtrivedi16 great work, thanks for the new pattern!

**Comment by iluwatar** [comment]
@all-contributors please add @ashishtrivedi16 for code

**Pull request merged**